src/Domain/CoordinateMaps/UniformCylindricalFlatEndcap.cpp

```diff
@@ -0,0 +1,712 @@
+      horizontal_dist_spheres + radius_one_ * sin(theta_max_one_) - radius_two_;
+
+  const double alpha =
+      atan2(center_two[2] - z_plane_one, max_horizontal_dist_between_circles);
+  ASSERT(alpha > 1.1 * theta_max_one_,
+         "Angle alpha is too small: alpha = "
+             << alpha << ", theta_max_one = " << theta_max_one_
+             << ", max_horizontal_dist_between_circles = "
+             << max_horizontal_dist_between_circles
+             << ", horizontal_dist_spheres = " << horizontal_dist_spheres
+             << param_string());
+#endif
+}
+
+template <typename T>
+std::array<tt::remove_cvref_wrap_t<T>, 3>
+UniformCylindricalFlatEndcap::operator()(
+    const std::array<T, 3>& source_coords) const {
+  using ReturnType = tt::remove_cvref_wrap_t<T>;
+  const ReturnType& xbar = source_coords[0];
+  const ReturnType& ybar = source_coords[1];
+  const ReturnType& zbar = source_coords[2];
+  std::array<ReturnType, 3> target_coords{};
+  ReturnType& x = target_coords[0];
+  ReturnType& y = target_coords[1];
+  ReturnType& z = target_coords[2];
+
+  // Use y and z as temporary storage to avoid allocations,
+  // before setting them to their actual values.
+  z = sqrt(square(xbar) + square(ybar));  // rhobar in the dox
+  y = radius_one_ *
+          cylindrical_endcap_helpers::sin_ax_over_x(z, theta_max_one_) *
+          (1.0 - zbar) +
+      radius_two_ * (1.0 + zbar);
+
+  x = 0.5 * (y * xbar + center_one_[0] * (1.0 - zbar) +
+             center_two_[0] * (1.0 + zbar));
+  y = 0.5 * (y * ybar + center_one_[1] * (1.0 - zbar) +
+             center_two_[1] * (1.0 + zbar));
+  z = 0.5 *
+      ((radius_one_ * cos(theta_max_one_ * z) + center_one_[2]) * (1.0 - zbar) +
+       center_two_[2] * (1.0 + zbar));
+  return target_coords;
+}
+
+std::optional<std::array<double, 3>> UniformCylindricalFlatEndcap::inverse(
+    const std::array<double, 3>& target_coords) const {
+  // First do some easy checks that target_coords is in the range
+  // of the map.  Need to accept points that are out of range by roundoff.
+
+  // check z >= z_plane_one_
+  if (target_coords[2] < z_plane_one_ and
+      not equal_within_roundoff(target_coords[2], z_plane_one_)) {
+    return std::nullopt;
+  }
+
+  // check that target point is at smaller z than circle_two
+  if (target_coords[2] > center_two_[2] and
+      not equal_within_roundoff(target_coords[2], center_two_[2])) {
+    return std::nullopt;
+  }
+
+  // check that point is outside or on sphere_one_
+  const double r_minus_center_one_squared =
+      square(target_coords[0] - center_one_[0]) +
+      square(target_coords[1] - center_one_[1]) +
+      square(target_coords[2] - center_one_[2]);
+  if (r_minus_center_one_squared < square(radius_one_) and
+      not equal_within_roundoff(sqrt(r_minus_center_one_squared),
+                                radius_one_)) {
+    // sqrt above because we don't want to change the scale of
+    // equal_within_roundoff too much.
+    return {};
```

Comment:
What's the difference between `return {}` and return `std::nullopt`, if any? If none, why not be consistent?

Reply:
Oops. Those should have been std::nullopt.  Changed.

(there is no difference.  `return {}` and `return std::nullopt` both construct an invalid std::optional).